Reimplementation of bridge/multifile feature for data_override in main

nikizadehgfdl opened the pull request:
- Should replace PR #1214 
- Raphael Dussin implementation of bridge/multifile feature for data_override
- This commit is basically Raph's branch raphaeldussin:feature_data_override_multifile merged with the main branch of FMS in PR #1214
- Due to restructure of time_interp and data_override dirs in FMS2023.03 I could not (did not know how to) do this merge via git and had to extract the mods from that PR and apply them to the main branch.
- Does not change answers.

**Description**
This updates enables using multiple data source files for data_override for one field. This is necessary for example when the data is contained in annual files (JRA or CORE datasets) and model time is at the start or end of one model year. Data_override needs the data at the end or start of the adjacent year to interpolate. Currently this is a limitation of data_override and the source data have to be specially prepared for GFDL models by a process called padding (prepend the last few time levels of the previous year and append the first few levels of the next year). 
This updates enables data_override to work without such padding by specifying multiple files in the data_table like:
```
"ATM", "p_surf",  "psl", "./INPUT/JRA_psl_ym1.nc:./INPUT/JRA_psl_yy1.nc:./INPUT/JRA_psl_yp1.nc",     "bilinear",   1.0

```
where JRA_psl_ym1.nc is data for previous model year
JRA_psl_yy1.nc   is data for current model year
JRA_psl_yp1.nc is data for next model year

**How Has This Been Tested?**
This has been tested in ocean-ice models with JRA dataset and reproduces same answers.

**Checklist:**
- [* ] My code follows the style guidelines of this project
- [ *] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ *] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] New check tests, if applicable, are included
- [ ] `make distcheck` passes

nikizadehgfdl commented:
@raphaeldussin this is what I have for the "merge" of your branch to main. I made this PR just so you and others could review what is involved, how you use it is up to you.  
Because of the time_interp and data_override refactoring in FMS2023.03 I don't know how to compare this branch with yours via git. However, it yields the same answers for your  OM4p25_ERA5_spinup_v4 experiments as your branch and does not change answers or timings for OMIP_JRA.

nikizadehgfdl commented:
This PR did not have a merge conflict yesterday! Is  the conflict because of the keywords change for "data" ?

raphaeldussin commented:
thanks a lot @nikizadehgfdl ! I'll have a look tonight and fix the conflicts.

uramirez8707 commented:
@raphaeldussin Thanks for implementing this and thanks @nikizadehgfdl for solving the merge conflicts (sorry you have to do it again)

The code updates look good to me. 
@raphaeldussin Do you get any difference in performance between using this new bridging method and the regular method? Particularly at initialization?

raphaeldussin commented:
Regarding the performance at initialization, I found no extra cost. I've shared with @nikizadehgfdl my xml and job output so he could confirm my findings.

bensonr reviewed (commented):
I may have missed it, but Is it required that all files in the multi-file list have the same spatial dimensions?  If it is required, is there a check to ensure?

raphaeldussin commented on data_override/include/data_override.inc:
```diff
@@ -441,6 +468,20 @@ subroutine read_table(data_table)
+            data_entry%multifile = .true.
+            index_1col = index(data_entry%file_name, ":")
+            index_2col = index(data_entry%file_name(index_1col+1:), ":")
+            if (index_2col .eq. 0) call mpp_error(FATAL, "data_override: must provide either 1 or 3 forcing files")
```
2 files poses the question of "should it go before or after?". Also most of the time we need to provide a duplicate of the very first record of the timeserie for time interpolation (unless first record falls exactly at midnight of Jan 1st) at the beginning of the run (and similar for the end) so we always end up padding at the beginning and end of all years

raphaeldussin commented on data_override/include/data_override.inc:
```diff
@@ -399,6 +412,20 @@ subroutine read_table(data_table)
                                  & region_type
+
+             if (index(data_entry%file_name, ":") .ne. 0) then
+               data_entry%multifile = .true.
+               index_1col = index(data_entry%file_name, ":")
+               index_2col = index(data_entry%file_name(index_1col+1:), ":")
+               data_entry%prev_file_name = data_entry%file_name(1:index_1col-1)
+               data_entry%next_file_name = data_entry%file_name(index_1col+index_2col+1:)
+               ! once previous/next files are filled in, overwrite current
+               data_entry%file_name = data_entry%file_name(index_1col+1:index_1col+index_2col-1)
```
we could test for 2 occurences of  the `:` separator between files to assure there are either 1 or 3 files but not 2. This would bulletproof the code I agree, good catch!

nikizadehgfdl commented:
> Regarding the performance at initialization, I found no extra cost. I've shared with @nikizadehgfdl my xml and job output so he could confirm my findings.

I looked at the clocks for the same 1 month ocean-ice experiment forced with JRA data with and without the multi-file-bridge data_override (apples to apples). The Ocean and Ice performance were the same. The nice surprise was that  the data-override/flux-exchange actually benefited from the multi-file scheme and the whole model **sped up** by ~6%. 
```
clock              multi-file(secs)     single-file(secs)       
Main loop                      857   911
  A-L: sfc_boundary_layer       24    62
  A-L: flux_down_from_atmos     20    42
```
I don't fully understand this since the test was done for the month of October which only one forcing file is involved.

nikizadehgfdl commented:
I am done with the merging and fixing the CI tests. The one that fails (Test coupler build) has nothing to do with this PR. 
@raphaeldussin could you take it from here addressing the reviewers comments?  We'd better push this through while there is no merge conflict. Thanks.

raphaeldussin commented:
@nikizadehgfdl many thanks for this! What is the preferred method here? I can work off your branch and PR to it if that's ok with you? Then this PR can be merged in main and we abandon my original PR. sounds good to you?

uramirez8707 commented:
@raphaeldussin I implemented this in the yaml version of the data table and updated the documentation in https://github.com/nikizadehgfdl/FMS/pull/3

Let me know if you are okay with it or would like any changes. 

I think this is the last thing left to do.

thomas-robinson commented on data_override/include/data_override.inc:
```diff
@@ -1107,48 +1362,228 @@ subroutine DATA_OVERRIDE_3D_(gridname,fieldname_code,return_data,time,override,d
+          endif
+          call time_interp_external(id_time,time,return_data(startingi:endingi,startingj:endingj,1),verbose=.false., &
+                                    is_in=is_in,ie_in=ie_in,js_in=js_in,je_in=je_in,window_id=window_id)
+        endif  ! end bridge
```
This end bridge comment is confusing because there are two if blocks documented at bridging

thomas-robinson commented on data_override/include/data_override.inc:
```diff
@@ -1107,48 +1362,228 @@ subroutine DATA_OVERRIDE_3D_(gridname,fieldname_code,return_data,time,override,d
-                 horz_interp=override_array(curr_position)%horz_interp(window_id), &
-                 is_in=is_in,ie_in=ie_in,js_in=js_in,je_in=je_in,window_id=window_id)
+
+           if (multifile) then   ! bridging between files, see previous blocks for comments
```
The previous blocks don't really have comments.

thomas-robinson commented on data_override/include/data_override.inc:
```diff
@@ -1107,48 +1362,228 @@ subroutine DATA_OVERRIDE_3D_(gridname,fieldname_code,return_data,time,override,d
         !! size
-        call time_interp_external(id_time,time,return_data(startingi:endingi,startingj:endingj,:),verbose=.false., &
-                               is_in=is_in,ie_in=ie_in,js_in=js_in,je_in=je_in,window_id=window_id)
+        if (multifile) then  ! bridging between files, see previous blocks for comments
```
This if statement should be labeled and documented

thomas-robinson commented on data_override/include/data_override.inc:
```diff
@@ -1107,48 +1362,228 @@ subroutine DATA_OVERRIDE_3D_(gridname,fieldname_code,return_data,time,override,d
-        call time_interp_external(id_time,time,return_data(startingi:endingi,startingj:endingj,:),verbose=.false., &
-                               is_in=is_in,ie_in=ie_in,js_in=js_in,je_in=je_in,window_id=window_id)
+        if (multifile) then  ! bridging between files, see previous blocks for comments
+          if (time<first_record) then
```
This if statement should be labeled and documented

thomas-robinson commented on data_override/include/data_override.inc:
```diff
@@ -1107,48 +1362,228 @@ subroutine DATA_OVERRIDE_3D_(gridname,fieldname_code,return_data,time,override,d
-                 mask_out   =mask_out(:,:,1), &
-                 is_in=is_in,ie_in=ie_in,js_in=js_in,je_in=je_in,window_id=window_id)
+
+           if (multifile) then  ! bridging between files, see previous blocks for comments
```
This if statement should be labeled and documented

thomas-robinson commented on data_override/include/data_override.inc:
```diff
@@ -1161,17 +1596,78 @@ subroutine DATA_OVERRIDE_3D_(gridname,fieldname_code,return_data,time,override,d
-              is_in=is_in,ie_in=ie_in,js_in=js_in,je_in=je_in,window_id=window_id)
-              return_data = return_data*factor
+
+           if (multifile) then  ! bridging between files, see previous blocks for comments
```
This if statement should be labeled and documented

thomas-robinson commented on time_interp/include/time_interp_external2_bridge.inc:
```diff
@@ -0,0 +1,295 @@
+    subroutine TIME_INTERP_EXTERNAL_BRIDGE_2D_(index1, index2, time, data_in, interp, verbose,horz_interp, mask_out, &
+               is_in, ie_in, js_in, je_in, window_id)
+
+      integer, intent(in) :: index1, index2
+      type(time_type), intent(in) :: time
+      real(FMS_TI_KIND_), dimension(:,:), intent(inout) :: data_in
+      integer, intent(in), optional :: interp
+      logical, intent(in), optional :: verbose
+      type(horiz_interp_type),intent(in), optional :: horz_interp
+      logical, dimension(:,:), intent(out), optional :: mask_out !< set to true where output data is valid
+      integer,                  intent(in), optional :: is_in, ie_in, js_in, je_in
+      integer,                  intent(in), optional :: window_id
+
+      real(FMS_TI_KIND_), dimension(size(data_in,1), size(data_in,2), 1) :: data_out
+      logical, dimension(size(data_in,1), size(data_in,2), 1) :: mask3d
```
This is a new subroutine, so these variables should all be documented in doxygen-style

thomas-robinson commented on time_interp/include/time_interp_external2_bridge.inc:
```diff
@@ -0,0 +1,295 @@
+    subroutine TIME_INTERP_EXTERNAL_BRIDGE_3D_(index1, index2, time, time_data, interp,verbose,horz_interp, mask_out, &
+                                             &  is_in, ie_in, js_in, je_in, window_id)
+
+      integer,                    intent(in) :: index1,index2 !< index of external field from previous call
+                                                                            !! to init_external_field
+      type(time_type),            intent(in) :: time !< target time for data
+      real(FMS_TI_KIND_), dimension(:,:,:),  intent(inout) :: time_data !< global or local data array
+      integer,                    intent(in), optional :: interp
+      logical,                    intent(in), optional :: verbose !< flag for debugging
+      type(horiz_interp_type),    intent(in), optional :: horz_interp
+      logical, dimension(:,:,:), intent(out), optional :: mask_out !< set to true where output data is valid
+      integer,                    intent(in), optional :: is_in, ie_in, js_in, je_in
+      integer,                    intent(in), optional :: window_id
+
+      type(time_type) :: time1, time2
+      integer :: dims1(4), dims2(4)
+      integer :: first_rec, last_rec
+      integer :: nx, ny, nz, interp_method, t1, t2
+      integer :: i1, i2, isc, iec, jsc, jec, mod_time
+      integer :: isc1, iec1, jsc1, jec1, isc2, iec2, jsc2, jec2
+      integer :: yy, mm, dd, hh, min, ss
+      character(len=256) :: err_msg, filename
+
+      integer :: isw, iew, jsw, jew, nxw, nyw
+          ! these are boundaries of the updated portion of the "data" argument
+          ! they are calculated using sizes of the "data" and isc,iec,jsc,jsc
+          ! fileds from respective input field, to center the updated portion
+          ! in the output array
+
+      real(FMS_TI_KIND_) :: w1,w2
+      logical :: verb
+      character(len=16) :: message1, message2
+      integer, parameter :: kindl = FMS_TI_KIND_
```
These variables need doxygen-style documentation

thomas-robinson commented on time_interp/include/time_interp_external2_bridge.inc:
```diff
@@ -0,0 +1,295 @@
+
+    subroutine TIME_INTERP_EXTERNAL_BRIDGE_0D_(index1, index2, time, time_data, verbose)
+
+      integer, intent(in) :: index1, index2
+      type(time_type), intent(in) :: time
+      real(FMS_TI_KIND_), intent(inout) :: time_data
+      logical, intent(in), optional :: verbose
+
+      integer :: t1, t2
+      integer :: i1, i2, mod_time
+      integer :: yy, mm, dd, hh, min, ss
+      character(len=256) :: err_msg, filename
+      type(time_type) :: time1, time2
+      integer :: dims1(4), dims2(4)
+
+      real(FMS_TI_KIND_) :: w1,w2
+      logical :: verb
+      integer, parameter :: kindl = FMS_TI_KIND_
```
These variables need doxygen-style documentation

thomas-robinson reviewed (changes requested):
1.  Any new variables need to be documented with doxygen-style comments
2.  There are a lot of new big `if` blocks in data_override.inc.  I stopped adding the comment that they should be documented and labeled because there are so many.  Please label the large if blocks to make reading the code easier in the future and add comments to explain what the if block is doing.  I think it would also be helpful to add to the function/subroutine descriptions
3. Is it possible to put the `if` blocks into a function and then call the function?  Are these if blocks copy-and-pastes?  They seem similar, and a function or subroutine would help reduce the size of this code.
4. I may have missed it, but I don't think that the term "bridge" is ever defined or discussed anywhere in the code documentation.  "Bridge" is used extensively through the documentation, so I think there needs to be a clear, longer explaination of what is happening and what "bridge" means.

mlee03 commented on time_interp/include/time_interp_external2_bridge.inc:
```diff
@@ -0,0 +1,295 @@
+      time1 = loaded_fields(index1)%time(t1)
+      time2 = loaded_fields(index2)%time(t2)
+      w2= (time - time1) // (time2 - time1)
+      w1 = 1.0-w2
```
1.0_kindl

mlee03 commented on time_interp/include/time_interp_external2_bridge.inc:
```diff
@@ -0,0 +1,295 @@
+      time1 = loaded_fields(index1)%time(t1)
+      time2 = loaded_fields(index2)%time(t2)
+      w2= (time - time1) // (time2 - time1)
+      w1 = 1.0-w2
```
1.0_kindl

mlee03 commented on time_interp/include/time_interp_external2_bridge.inc:
```diff
@@ -0,0 +1,295 @@
+
+      time1 = loaded_fields(index1)%time(t1)
+      time2 = loaded_fields(index2)%time(t2)
+      w2= (time - time1) // (time2 - time1)
```
got it got it got it, the "//" operator.  It's grayed out on Github and I immediately took it for a comment without thinking :) .  All's good.

raphaeldussin commented:
@thomas-robinson @mlee03 I think the changes address most of your concerns. Additionally:

* I agree that the number of `if` blocks is not ideal. One could move this into a new function but since each block calls `time_interp` with different set of optional arguments there is little we could do to reduce the code bloat.
* I have added more comments and a better explanation about the "bridging" process in data_override. I'm happy to add some more documentation now or later if you can point me to the appropriate place to do so.

thomas-robinson reviewed (approved):
Thank you for your updates.  I think this code is much easier to read now.

nikizadehgfdl commented on data_override/include/data_override.inc:
```diff
@@ -695,8 +784,71 @@ subroutine DATA_OVERRIDE_0D_(gridname,fieldname_code,data_out,time,override,data
+    endif if_time2
+  else ! standard behavior
+     if ((time<first_record) .or. (time>last_record)) then
+       call mpp_error(WARNING, &
```
This WARNING is swamping the stdout for regression tests that were/are working fine. There is something wrong with this logic.
